### PR TITLE
Fixing type propagation behaviour between `NewExpression` and `ConstructExpression`

### LINKED ISSUE #493: Stackoverflow in `ParserType::hashCode`
The concurrent frontends (#444) somehow managed to introduce a stack overflow into the Type's hashcode function.

```
Caused by: java.lang.StackOverflowError
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.Type.hashCode(Type.java:357)
	at de.fraunhofer.aisec.cpg.graph.types.ObjectType.hashCode(ObjectType.java:189)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
	at java.base/java.util.Arrays.hashCode(Arrays.java:4685)
	at java.base/java.util.Objects.hash(Objects.java:146)
	at de.fraunhofer.aisec.cpg.graph.types.PointerType.hashCode(PointerType.java:159)
```

See https://github.com/Fraunhofer-AISEC/codyze/runs/3232025152?check_suite_focus=true

### PR DESCRIPTION
**Pull request by oxisto**
Type propagation between `NewExpression` and `ConstructExpression` were broken. Partly introduced by #444, partly before. This PR tries to fix this with more constant behaviour. Basically, the language frontends are now in charge of properly creating the types of `NewExpression`s and `ConstructExpression`. Before that, it somehow relied on a weird mix of type propagation and a `fixInitializer` function. Now I clearly documented in the wiki, that the language frontend needs to take care of this. This is needed because the behaviour is different between languages with pointer and non-pointer types.

In an ideal cpg, we would have type propagation rules that could be configured per language frontend.

There is still a (weird) type propagation *towards* `ConstructExpression`, mainly from variable declarations. I will leave this for now, even though there should never be any type propagation *towards* a construct expression, but only *from* it.

This fixes #493

### CODE REVIEW

File: cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ConstructExpression.kt [hunk 1]
```diff
@@ -98,7 +88,7 @@ class ConstructExpression : CallExpression(), HasType.TypeListener {
         if (!TypeManager.isTypeSystemActive()) {
             return
         }
-        val subTypes: MutableSet<Type> = HashSet(getPossibleSubTypes())
+        /*val subTypes: MutableSet<Type> = HashSet(getPossibleSubTypes())
```
Comment (konradweiss) [inline cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ConstructExpression.kt:91]: Do we want to remove this commented out code or is there a readon for keeping it?

**Review by konradweiss — CHANGES_REQUESTED** [review]
I would remove the commented out code, unless there is a readon to keep it

File: cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ConstructExpression.kt (hunk 1 above)
Comment (oxisto) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ConstructExpression.kt:91]: It was intentionally left for the reviewer ;) Nah, it was left over code. I will remote it.

Comment (konradweiss) [reply cpg-library/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ConstructExpression.kt:91]: So i passed the reviewing the reviewer test ;)

**Review by konradweiss — APPROVED** [review]

**Pull request merged**